fpdcc/ccfp-asset-dashboard
# Fix phase / asset GIS bug

## Overview

- closes #224 
- closes #223 

includes:
- gis bug fix (#224)
- districts overwritten fix (#223)
- refactoring of the signals (see notes)
- tests (~half of the diff is a geojson test fixture)
- upgraded django to the latest patch version (see latest security release: https://www.djangoproject.com/weblog/2023/may/03/security-releases/)
- upgrade python (3.8 -> 3.9)

### Notes

#### gis bug fix
The comments in that issue include a lot of background about that bug, but TLDR: the intersects query was causing a segmentation fault on posgres when trying to find the intersection of political districts and a phase's assets. This happened sometimes, but not always. To fix, I'm filtering first with the `dwithin` function, then doing the intersection. The `dwithin` function trims down the area where we want to find the intersection. I've added a comment to this PR so you can see the dwithin change, since this diff is bigger with the refactor.

Update: I've upgraded postgis in the docker container so this is fixed on local development. We're migrating to CCFP's servers, so we'll hold off on fixing this on heroku (see this comment: https://github.com/fpdcc/ccfp-asset-dashboard/issues/224#issuecomment-1549877563)

#### signal refactor
Also includes a refactor of the django signal and GIS processing code. This fixes #223.

I've refactored all of the GIS logic into a class. I also found that a lot of the work was duplicated across signals, so I've consolidated all of the asset-related signals into one single signal so there's not as much redundant work ([see this comment in the bug issue about the inefficient, redundant code](https://github.com/fpdcc/ccfp-asset-dashboard/issues/224#issuecomment-1538948373)). This also makes all of the GIS processing logic easier to debug, follow, etc, since it's all in one place rather than split across signals.

## Testing Instructions
- visit this project on the review app: https://ccfp-asset-d-patch-sign-5yfa9e.herokuapp.com/projects/2/
  - login with temp credentials: username=`admin` && password=`password`
- in the project detail view linked above, scroll down and click "edit assets"
- search for an asset in cook county. bonus points if you find one that is geographically far away from the existing assets of that phase.
- add those assets to the phase. this action should show a success message.
- navigate back to the project overview (click the "overview" button). make sure the scores updated, specifically the political districts, geographic distance score, and social equity score
